### PR TITLE
Create TrainState dataclass

### PR DESCRIPTION
**Pull request by maanug-nv**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Creates a `TrainState` dataclass global variable for holding mutable training loop state, instead of holding it in `args`.
Part of #3555 .

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] If this PR adds or changes a GPU kernel (Triton, `jit_fuser`/`torch.compile`, CUDA extension, TE or external-library dispatch, or a scatter/index accumulation), I have added or updated its bit-exact determinism test and registered it in `tests/unit_tests/determinism/kernels/manifest.py` ([guide](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/developer/determinism/testing.md))
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/state.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        """Serializes the training state into a dictionary of tensors.
+
+        Conforms to the Stateful interface for distributed checkpointing.
+
+        Returns:
+            A dictionary where keys are state variable names and values are
+            their corresponding tensor representations.
+        """
+        return {
+            "step": torch.tensor(self.iteration, dtype=torch.int64),
```
Comment (cuichenx) [inline megatron/training/state.py:43]: let's maybe document why the discrepancy between "step" and "iteration"

File: megatron/training/state.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    Inherits from Stateful for distributed checkpointing compatibility.
+    Tracks iteration count, consumed samples, flags for train/valid/test phases,
+    and floating-point operations.
+    """
+
+    iteration: int = 0
+    consumed_train_samples: int = 0
+    skipped_train_samples: int = 0
+    consumed_valid_samples: int = 0
+    floating_point_operations_so_far: float = 0.0
```
Comment (cuichenx) [inline megatron/training/state.py]: maybe we can keep megatron's `num_floating_point_operations_so_far` name here to be consistent. Also should this be an integer?

**Review by cuichenx — APPROVED** [review]
LGTM

**Comment by chtruong814** [comment]
CI ran as part of a batch of PRs in the merge queue to consolidate CI runs
https://github.com/NVIDIA/Megatron-LM/pull/7462
https://github.com/NVIDIA/Megatron-LM/actions/runs/35306160723/job/105485956036

Tests passed, so will fast merge this.
